lib/wcgi-runner/src/builder.rs

```diff
@@ -0,0 +1,252 @@
+            ..self
+        }
+    }
+
+    /// Add an argument to the WASI executable's command-line arguments.
+    pub fn arg(mut self, arg: impl Into<String>) -> Self {
+        self.args.push(arg.into());
+        self
+    }
+
+    /// Add multiple arguments to the WASI executable's command-line arguments.
+    pub fn args<A, S>(mut self, args: A) -> Self
+    where
+        A: IntoIterator<Item = S>,
+        S: Into<String>,
+    {
+        self.args.extend(args.into_iter().map(|s| s.into()));
+        self
+    }
+
+    /// Expose an environment variable to the guest.
+    pub fn env(mut self, name: impl Into<String>, value: impl Into<String>) -> Self {
+        self.env.insert(name.into(), value.into());
+        self
+    }
+
+    /// Expose multiple environment variables to the guest.
+    pub fn envs<I, K, V>(mut self, variables: I) -> Self
+    where
+        I: IntoIterator<Item = (K, V)>,
+        K: Into<String>,
+        V: Into<String>,
+    {
+        self.env
+            .extend(variables.into_iter().map(|(k, v)| (k.into(), v.into())));
+        self
+    }
+
+    /// Forward all of the host's environment variables to the guest.
+    pub fn forward_host_env(self) -> Self {
+        Builder {
+            forward_host_env: true,
+            ..self
+        }
+    }
+
+    /// Override the CGI dialect.
+    pub fn cgi_dialect(self, dialect: CgiDialect) -> Self {
+        Builder {
+            dialect: Some(dialect),
+            ..self
+        }
+    }
+
+    /// Map `guest_dir` to a directory on the host.
+    pub fn map_dir(mut self, guest_dir: impl Into<String>, host_dir: impl Into<PathBuf>) -> Self {
+        self.mapped_dirs.push((guest_dir.into(), host_dir.into()));
+        self
+    }
+
+    /// Map one or more `guest_dir` to a directory on the host.
+    pub fn map_dirs<I, G, H>(mut self, mapping: I) -> Self
+    where
+        I: IntoIterator<Item = (G, H)>,
+        G: Into<String>,
+        H: Into<PathBuf>,
+    {
+        for (guest_dir, host_dir) in mapping {
+            self.mapped_dirs.push((guest_dir.into(), host_dir.into()));
+        }
+
+        self
+    }
+
+    /// Pass in a [`Handle`] to a custom tokio runtime.
+    pub fn tokio_handle(self, handle: Handle) -> Self {
+        Builder {
+            tokio_handle: Some(handle),
+            ..self
+        }
+    }
+
+    /// Set the engine used to compile the WebAssembly module.
+    pub fn engine(self, engine: Engine) -> Self {
+        Builder {
+            engine: Some(engine),
+            ..self
+        }
+    }
+
+    /// Create a [`Runner`] that executes a WEBC file.
+    ///
+    /// If [`Builder::program`] was set, this will look for the command with
+    /// that name in the WEBC file. Otherwise, it will fall back to the WEBC
+    /// file's default entrypoint.
+    ///
+    /// This will infer the [`CgiDialect`] from the WEBC file's metadata
+    pub fn build_webc(self, webc: impl Into<Bytes>) -> Result<Runner, Error> {
```

Comment:
I don't think `build_webc` is a good name. Probably `use_webc` or `from_webc` are way better names

Reply:
Also, you may want to provide a command to execute, rather than infer it. And get the webc directly instead of the bytes.

That way there are two steps:
1. parsing the webc
2. detecting the command to run and what kind it is
3. Run the runner

Reply:
I deliberately named it that way so we can follow the `Runner::builder().foo().bar().build()` pattern. Using "use" or "from" would be very confusing. 

In this case, we have multiple ways of constructing the final runner (e.g. a WEBC file, a WebAssembly binary, or the filename for something on disk), so there are multiple `build_xxx()` methods.

Reply:
I see the point that the `build_webc` actually returns a Runner (which I didn't realize before), so I kind of agree on that naming if we want to return a runner.

However, I think the webc shall be parsed beforehand and then chose the command to run.
For example, let's say you have a package with two commands:
* A (WASI command)
* B (WCGI command)

You don't know which runner you are going to run, until you have parsed the webc package. And at that point you already have the config and such. So passing a stream of bytes to parse again the webc seems wrong